File: apps/backend/internal/model/post/post.go
package post

import (
	"time"

	"github.com/google/uuid"
	"github.com/sarbojitrana/nexus/internal/model"
)

type PostType string

const (
	PostTypeComment PostType = "comment"
	PostTypePost    PostType = "post"
)

type VoteType int16

const (
	Upvote   VoteType = 1
	DownVote VoteType = -1
)

type Post struct {
	model.Base
	AuthorID     uuid.UUID  `json:"authorId" db:"author_id"`
	CommunityID  *uuid.UUID `json:"communityId" db:"community_id"`
	ParentPostID *uuid.UUID `json:"parentPostId" db:"parent_post_id"`
	PostType     PostType   `json:"postType" db:"post_type"`
	Title        *string    `json:"title" db:"title"`
	Content      *string    `json:"content" db:"content"`
	Upvotes      int        `json:"upvotes" db:"upvotes"`
	Downvotes    int        `json:"downvotes" db:"downvotes"`
	CommentCount int        `json:"commentCount" db:"comment_count"`
	DeletedAt    *time.Time `json:"deletedAt" db:"deleted_at"`
}

type PostMedia struct {
	model.BaseWithId
	UserID      uuid.UUID `json:"userId" db:"user_id"`
	PostID      uuid.UUID `json:"postId" db:"post_id"`
	DownloadKey string    `json:"downloadKey" db:"download_key"`
	FileSize    int64     `json:"fileSize" db:"file_size"`
	MimeType    string    `json:"mimeType" db:"mime_type"`
	model.BaseWithCreatedAt
}

type PostVote struct {
	PostID   uuid.UUID `json:"postId" db:"post_id"`
	UserID   uuid.UUID `json:"userId" db:"user_id"`
	VoteType VoteType  `json:"voteType" db:"vote_type"`
	model.BaseWithCreatedAt
	model.BaseWithUpdatedAt
}

type PopulatedPost struct {
	Post
	PostMedia []PostMedia `json:"postMedia" db:"post_media"`
}

File: packages/zod/src/post.ts
import { z } from "zod";
import { ZOrder, schemaWithCursorPagination, schemaWithPagination } from "@/utils.js";

export const ZPostType = z.enum(["comment", "post"]);
export const ZVoteType = z.enum(["upvote", "downvote"]);

export const ZPostMedia = z.object({
  id: z.string().uuid(),
  userId: z.string().uuid(),
  postId: z.string().uuid(),
  downloadKey: z.string(),
  fileSize: z.number(),
  mimeType: z.string(),
  createdAt: z.string().datetime(),
});

export const ZPost = z.object({
  id: z.string().uuid(),
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
  authorId: z.string(),
  communityId: z.string().uuid().nullable(),
  parentPostId: z.string().uuid().nullable(),
  postType: ZPostType,
  title: z.string().nullable(),
  content: z.string().nullable(),
  upvotes: z.number(),
  downvotes: z.number(),
  commentCount: z.number(),
  engagement: z.number(),
  isPopularityUpdated: z.boolean(),
  deletedAt: z.string().datetime().nullable(),
});

export const ZPopulatedPost = ZPost.extend({
  postMedia: z.array(ZPostMedia),
});

// parentPostId requires postType "comment" and vice versa
export const ZCreatePostPayload = z.object({
  communityId: z.string().uuid().nullable().optional(),
  parentPostId: z.string().uuid().nullable().optional(),
  postType: ZPostType,
  title: z.string().nullable().optional(),
  content: z.string().nullable().optional(),
});

export const ZUpdatePostByIDPayload = z.object({
  title: z.string().nullable().optional(),
  content: z.string().nullable().optional(),
});

export const ZGetCommentsByPostIDQuery = z.object({
  cursorSortValue: z.string().optional(),
  cursorCreatedAt: z.string().datetime().optional(),
  sort: z.enum(["created_at", "popularity"]).optional(),
  order: ZOrder.optional(),
  dateCreatedStart: z.string().datetime().optional(),
  dateCreatedEnd: z.string().datetime().optional(),
});

export const ZGetRepliesByCommentIDQuery = z.object({
  page: z.coerce.number().min(1).optional(),
  limit: z.coerce.number().min(1).max(100).optional(),
});

export const ZReactToPostPayload = z.object({
  reaction: ZVoteType,
});

export const ZGetPostsQuery = z.object({
  referenceTime: z.string().datetime().optional(),
  trendingLimit: z.coerce.number().optional(),
  followingUsersLimit: z.coerce.number().optional(),
  followingCommunitiesLimit: z.coerce.number().optional(),
  trendingCursorValue: z.coerce.number().optional(),
  trendingCursorCreatedAt: z.string().datetime().optional(),
  followingUsersCursorCreatedAt: z.string().datetime().optional(),
  followingCommunitiesCursorCreatedAt: z.string().datetime().optional(),
});

export const ZGetPostsQueryResponse = z.object({
  referenceTime: z.string().datetime(),

  trendingPosts: z.array(ZPopulatedPost),
  nextTrendingCursorValue: z.number().nullable(),
  nextTrendingCursorCreatedAt: z.string().datetime().nullable(),
  hasMoreTrending: z.boolean(),

  followingUsersPosts: z.array(ZPopulatedPost),
  nextFollowingUsersCursorCreatedAt: z.string().datetime().nullable(),
  hasMoreFollowingUsers: z.boolean(),

  followingCommunitiesPosts: z.array(ZPopulatedPost),
  nextFollowingCommunitiesCursorCreatedAt: z.string().datetime().nullable(),
  hasMoreFollowingCommunities: z.boolean(),
});

export const ZPostCommentsResponse = schemaWithCursorPagination(ZPopulatedPost);
export const ZPostRepliesResponse = schemaWithPagination(ZPopulatedPost);
